Lib/tarfile.py

```diff
@@ -533,26 +530,31 @@ def _read(self, size):
         if self.comptype == "tar":
             return self.__read(size)
 
-        c = len(self.dbuf)
-        t = [self.dbuf]
+        c = 0
+        t = []
         while c < size:
             # Skip underlying buffer to avoid unaligned double buffering.
             if self.buf:
                 buf = self.buf
                 self.buf = b""
+            elif self.comptype != "gz" and not self.cmp.needs_input:
+                buf = b""
             else:
                 buf = self.fileobj.read(self.bufsize)
                 if not buf:
                     break
             try:
-                buf = self.cmp.decompress(buf)
+                buf = self.cmp.decompress(buf, size - c)
+                if self.comptype == "gz":
+                    self.buf = self.cmp.unconsumed_tail
             except self.exception as e:
                 raise ReadError("invalid compressed data") from e
             t.append(buf)
             c += len(buf)
         t = b"".join(t)
-        self.dbuf = t[size:]
-        return t[:size]
+        if len(t) > size:
+            raise ReadError("decompress() returned too much data")
```

Comment:
Do you see any scenario where this would be triggered? Looking at the zlib, bz2 and lzma decompressor docs for max_length, it looks like this shouldn't occur?

I have no issue with it being here but just checking if I'm missing something :)

Reply:
Right - it can only happen if there is a bug in the zlib, bz2 or lzma decompressor. I haven't checked their C source, but their docs say it should not occur.
It's not a necessary check, but I figured I'd add it just in case.